Data/Aeson/Types/FromJSON.hs

```diff
@@ -952,18 +952,15 @@ type ConName = String
 contextType :: TypeName -> Parser a -> Parser a
 contextType = prependContext
 
-
-
 -- | Add the tagKey that will be looked up while building an ADT
 -- | Produce the error equivalent to
 -- | Left "Error in $: parsing T failed, expected an object with keys "tag" and
 -- | "contents", where "tag" i-- |s associated to one of ["Foo", "Bar"],
 -- | The parser returned error was: could not find key "tag"
 contextTag :: Text -> [String] -> Parser a -> Parser a
-contextTag tagKey cnames = prependFailure
-  ("expected an object with key \"" ++ unpack tagKey ++
-    "\" where \"tag\" is associated to one of " ++ show cnames ++
-    ". The parser returned error was: ")
+contextTag tagKey cnames = modifyFailure (\_ ->
+  ("expected Object with key \"" ++ unpack tagKey ++ "\"" ++
+    " associated to one of " ++ show cnames ++ "."))
```

Comment:
`prependFailure` would still be good here, only with a single comma after this new prefix.

Reply:
`prependFailure` would add the `key 'tag' not present' phrase here, which may be redundant, I think. That is the reason I removed it.

Reply:
I think error messages should be explicit about both what was expected and what was not expected, rather than let users guess the source of the mismatch, unless it's really obvious, which doesn't seem to be the case here. In particular, another potential source of error at this location would be if the `"key"` tag exists, but is associated to some non-string value.

What I was complaining about earlier was specifically the extra string `"the parser returned error was"`, because it does not seem like a user can easily understand what "the parser" refers to; it could refer to the "parser" `obj .: tagkey`, which reveals too much internal details that users may not be aware of, or it could refer to the toplevel function they are calling (`decode`, `parseJSON`...), which does make that part of the message redundant.

Reply:
Sure. I understand the we had to provide the original information to the user. And that is the reason I had added used 'prependFailure' earlier.  I can remove the 'the parser error was' as you suggested. But, in this context I found the last part of message 'key 'tag' not present' to be repetitive. If you think otherwise,  I can leave it there.

Reply:
`obj .: tagKey` does two things

- it looks up `tagKey` in `obj`, which fails if the key is not there, that's the first possible source of error, and the cause of the issue this PR is meant to fix.

- and it parses the associated value according to the `FromJSON` instance of the expected result's type, which in this case is `Text` (or `String` I'm not sure). This parsing can also fail, that's the second source of error that can happen at this point of the program.

Reply:
@Lysxia Thanks for the info. Here is now the message might look with `prependFailure` included.

```
>data T = Foo Int | Bar String deriving (Generic, FromJSON, ToJSON, Show)
>eitherDecode "{\"tag1\": \"Foo1\", \"contents\": 10}" ::  Either String T
Left "Error in $: parsing Ghci7.T failed, expected Object with key \"tag\" associated to one of [\"Foo\",\"Bar\"].key \"tag\" not present"
>eitherDecode "{\"tag\": 10, \"contents\": 10}" ::  Either String T
Left "Error in $.tag: parsing Ghci7.T failed, expected Object with key \"tag\" associated to one of [\"Foo\",\"Bar\"].parsing Text failed, expected String, but encountered Number"
```

You see that the second part of message in both scenarios are not Capitalized. Should we do that? Or is is common across all error message in aeson? Also, to my point the second part of the first error message 'key \"tag\" not present" is kind of redundant, and hence my previous change.

I think the current solution is to leave 'prependFailure' intact and let the redundant message be as shown above.

Reply:
Also, we have other places where we use 'not present'. Changing 'not present' to 'not found' might preclude changing those places as well.

Reply:
The "convention" is to prepend using a comma as a separator. It's just what I came up with when I refactored the error handling in this module, so don't hesitate to discuss it if you have better ideas!

I don't think the second part of the error message is redundant. If you say `expected Object with key \"tag\" associated to one of [\"Foo\",\"Bar\"].`, then there are two ways to break that expectation:

- the key `"tag"` doesn't exist
- or the key tag is not associated to one of these strings: it could be a different string (that is handled later on in the `do` block), or not a string at all.

So that's an extra bit of information you can give directly, rather than expect users from deducing it themselves in an error prone way.

---

I think "not found" is better than "not present" in all the other places too.